presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp

```diff
@@ -144,25 +144,32 @@ void PrestoExchangeSource::processDataResponse(
 
   std::unique_ptr<exec::SerializedPage> page;
   std::unique_ptr<folly::IOBuf> singleChain;
-  bool empty = response->empty();
+  const bool empty = response->empty();
+  int64_t totalBytes{0};
   if (!empty) {
     auto iobufs = response->consumeBody();
     for (auto& buf : iobufs) {
+      totalBytes += buf->capacity();
       if (!singleChain) {
         singleChain = std::move(buf);
       } else {
         singleChain->prev()->appendChain(std::move(buf));
       }
     }
+    PrestoExchangeSource::updateMemoryUsage(totalBytes);
```

Comment:
I feel having static variables is a bit hacky. ExchangeQueue is technically owned by ExchangeClient instead of ExchangeSource. It is passed in as a shared_ptr to ExchangeSource upon creation to let ExchangeSource enqueue pages. ExchangeSource comes and goes so it does not own the memory. So I think it is a bit weird to have the memory counter in ExchangeSource. Can we move it to ExchangeClient? Not ExchangeQueue because the SerializedPage inside ExchangeQueue will live a bit longer outside the queue after dequeuing.

Reply:
I think stats we collect here is the queuing at presto exchange source. The client might not be the good place as it is the consumer but not producer. The buffer is actually stayed in queue but we can't collect stats in queue because exchange queue has been used in many places not just for presto exchange source.

Reply:
exchange source is the producer of elements in the queue. Exchange client is the entity that owns the queue. Exchange operator is the consumer that consumes from the queue.